### PR TITLE
Make conversion of double to BigDecimal exact

### LINKED ISSUE #9607: Make `Float.to Decimal` conversion exact and hash `double` equivalent of a number
In our current implementation, `Integer` / `Float` pairs that are considered equal by `==` do not always hash the same. Values near `+/-Double.MAX_VALUE` in particular have different hash values. In contrast, `2` and `2.0` do have the same hash value, since our hash implementation takes care to make these hash the same, in most cases.

For consistency, we should make `Integer` and `Float` values hash differently, so that users do not come to the conclusion. This also makes Enso consistent with Java in this behavior.

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Resolves #9607 by computing `Number.hash` by converting given number to `Float` first and then computing the hash. Also the conversion from `Float.to Decimal` is exact - done via `new BigDecimal(double)`. There is `Decimal.new` that handles the user-friendly conversion. However as a result `Decimal.from 2.1 != Decimal.new 2.1` - that's the only way to ensure consistency between hash code and conversions.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

File: engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java [hunk 1]
```diff
@@ -369,6 +369,10 @@ public List<Value> numbers() {
     if (languages.contains(Language.ENSO)) {
       collect.add(v(null, "", "42").type());
       collect.add(v(null, "", "6.7").type());
+      collect.add(v(null, "", "9223372036854776000").type());
+      collect.add(v(null, "", "9223372036854775999").type());
+      collect.add(v(null, "", "9223372036854776000.0").type());
+      collect.add(v(null, "", "9223372036854775999.0").type());
```
Comment (JaroslavTulach) [inline engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: @GregoryTravis, please add here "all" the Enso numbers that you consider suspicious. Then execute `sbt runtime-integration-tests/testOnly *HashCodeTest`.

File: engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java [hunk 2]
```diff
@@ -392,6 +396,9 @@ public List<Value> numbers() {
       collect.add(ctx.asValue((short) 44));
       collect.add(ctx.asValue((int) 5432));
       collect.add(ctx.asValue((long) 5435432));
+      long over52bit = 2 ^ 56 + 777;
+      collect.add(ctx.asValue(over52bit));
+      collect.add(ctx.asValue((double) over52bit));
```
Comment (JaroslavTulach) [inline engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:403]: @GregoryTravis put here Java values that behave "wrongly". Then execute `sbt runtime-integration-tests/testOnly *HashCodeTest`.

File: .vscode/settings.json [hunk 3]
```diff
@@ -23,5 +23,7 @@
   ],
   "files.watcherExclude": {
     "**/target": true
-  }
+  },
+  "metals.inlayHints.implicitArguments.enable": true,
+  "metals.inlayHints.implicitConversions.enable": true
```
Comment (radeusgd) [inline .vscode/settings.json:28]: seems unrelated?

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsSimpleNode.java [hunk 4]
```diff
@@ -153,21 +154,32 @@ boolean equalsDoubleText(double self, Text other) {
...
     }
   }
 
   @TruffleBoundary
-  private static double asDouble(BigInteger big) {
-    return big.doubleValue();
+  private static boolean equalsDoubleBigInteger(double self, BigInteger big) {
+    var selfDecimal = new BigDecimal(self);
+    var otherDecimal = new BigDecimal(big);
+    return selfDecimal.equals(otherDecimal);
```
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsSimpleNode.java:182]: I ❤️ this, I think this is the right approach

**Review by radeusgd — APPROVED** [review]

File: .vscode/settings.json (hunk 3 above)
Comment (JaroslavTulach) [reply .vscode/settings.json:28]: Unrelated, but new version of `metals` insist on adding these options, so we can probably leave it in.

Comment (radeusgd) [reply .vscode/settings.json:28]: Ok, sure

**Comment by JaroslavTulach** [comment]
Eleven failures in standard libraries. Among others:
- [when finding a common type for numeric columns to be Float, any precision loss should be reported](https://github.com/enso-org/enso/actions/runs/8739985955/job/23982581834?pr=9740#step:7:6362)
- [should return decimals when rounding decimals](https://github.com/enso-org/enso/actions/runs/8739985955/job/23982581834?pr=9740#step:7:6541)
- [should allow to create a Float column from a big integer, but warn about Loss_Of_Integer_Precision if relevant](https://github.com/enso-org/enso/actions/runs/8739985955/job/23982581834?pr=9740#step:7:10736)
- [allows arithmetic on Decimal columns and other numeric columns](https://github.com/enso-org/enso/actions/runs/8739985955/job/23982581834?pr=9740#step:7:10740)

@GregoryTravis can you take a look and decided whether it is OK to update the tests? If so, feel free to push into this branch.

**Update**: there is also an [AssertionError](https://github.com/enso-org/enso/actions/runs/8749250691/job/24010598281?pr=9740#step:7:1209):
```
Reason: An unexpected panic was thrown: (Assertion_Error.Error 'java.lang.AssertionError: Different hash code! Original 1.7976931348623157E308[#318db821] 
got converted to WithWarnings{Value JavaObject[1.7976931348623157E+308 
(java.math.BigDecimal)] has 1 warnings}[#3c7d89b4]')
``` 
probably we need to support `iop.isBigInteger(...)` case in addition to `EnsoBigInteger` only check.

File: engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java (hunk 1 above)
Comment (GregoryTravis) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: On this branch, the only value that has ==/hash inconsistency is 9223372036854775808, which I have added.

Comment (GregoryTravis) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: When I run the tests on this branch:

```
sbt:enso> runtime-integration-tests/testOnly *HashCodeTest
[...]
[info] Passed: Total 0, Failed 0, Errors 0, Passed 0
[info] No tests to run for runtime-integration-tests / Test / testOnly
```

```
sbt:enso> project runtime-integration-tests
[info] set current project to runtime-integration-tests (in build file:/Users/gmt/dev/enso/enso/)
sbt:runtime-integration-tests> runtime-integration-tests/testOnly *HashCodeTest*
[...]
[info] Passed: Total 0, Failed 0, Errors 0, Passed 0
[info] No tests to run for Test / testOnly
```

Comment (JaroslavTulach) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: I am not sure what can be wrong? `sbt buildEngineDistribution` first, maybe?

Comment (JaroslavTulach) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: > On this branch, the only value that has ==/hash inconsistency is 9223372036854775808, which I have added.

Great example! `9223372036854775808 == 2^63` `BigDecimal` for both `double` and `BigInteger` are equal, but the hashcodes aren't.

The number is also `Long.MAX_VALUE + 1` where the constant is [0x7fffffffffffffffL](https://github.com/openjdk/jdk/blob/jdk-21%2B35/src/java.base/share/classes/java/lang/Long.java#L85)

Comment (JaroslavTulach) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: I believe the `2^63` value if fixed by 0413d5f and [96da44b](https://github.com/enso-org/enso/pull/9740/commits/96da44b0c35ea5501ee51216dd758770b268bf01) - the [other commit](https://github.com/enso-org/enso/pull/9740/commits/96da44b0c35ea5501ee51216dd758770b268bf01) uses `new BigDecimal(double)` instead of `valueOf` - they behave quite differently! Looks like the `new BigDecimal()` approach is the right one for the `2^63` case. We should review also other usages of `BigDecimal.valueOf`:

Usage of BigDecimal.valueOf

**Review by hubertp — APPROVED** [review]

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/number/utils/BigIntegerOps.java [hunk 5]
```diff
@@ -235,7 +235,12 @@ public static boolean fitsInLong(BigInteger bigInteger) {
   }
 
   public static boolean fitsInLong(double decimal) {
-    return decimal <= Long.MAX_VALUE && decimal >= Long.MIN_VALUE;
+    var nulaMinus = Double.doubleToRawLongBits(-0d);
+    if (nulaMinus == Double.doubleToRawLongBits(decimal)) {
+      return false;
+    }
```
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/number/utils/BigIntegerOps.java:245]: Does that mean that `-0.0` will no longer be equal to `0`?

I'd expect this to be true:
```
> 1/Number.positive_infinity
>>> 0.0
> -1/Number.positive_infinity
>>> -0.0
> (1/Number.positive_infinity) == (-1/Number.positive_infinity)
>>> true
> (-1/Number.positive_infinity) == 0
>>> true
```
(that's how it works currently)

I think we should add a test case to make sure it works as expected. In particular, we need to ensure we don't break transitivity of `==`: `0 == 0.0 == -0.0`

Comment (radeusgd) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/number/utils/BigIntegerOps.java:245]: I guess this will work because the `0` literal is `long`. But what if I create a `0` that is a `BigInteger`? It definitely must be consistent.

**Review by Akirathan — APPROVED** [review]

File: engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java (hunk 1 above)
Comment (GregoryTravis) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: @JaroslavTulach This is failing in `0.1.truncate`, since it uses `.toBigIntegerExact` in TruncateNode, and that throws an exception when the argument is not an integer.

Comment (GregoryTravis) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: Also, getting an inconsistent hash for `Float.max_value`:

```
    d = Decimal.new Float.max_value
    IO.println (d == Float.max_value)
```

```
Execution finished with an error: Assertion Error: 'java.lang.AssertionError: Different hash code! Original 1.7976931348623157E308[#318db821] got converted to WithWarnings{Value JavaObject[1.7976931348623157E
```

Comment (JaroslavTulach) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: I can see this is getting pretty complicated. To support `Decimal.from (that : Float)` conversion, we need to make sure the `hash` of original and new value is the same. That leads to `new BigInteger(that)`. However then `(double)0.1` is converted to a nasty number with fifteen+ decimal digits.

On the other hand, use of `BigInteger.valueOf` may lead to `Decimal` objects with different `hash` than the original one when doing `Decimal.from(that:Float)` conversion. My attempt to deal with it is visible in f39794796c7f185c732fc08247d187282a58a24d - let's see how well it goes with the tests.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/number/utils/BigIntegerOps.java (hunk 5 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/number/utils/BigIntegerOps.java:245]: I had to return this change back and rather do d69f62d where needed.

**Comment by JaroslavTulach** [comment]
> Eleven failures in standard libraries. Among others:

Now we have (again) eleven [failing tests in Table_Tests](https://github.com/enso-org/enso/actions/runs/8764871354/job/24055335007?pr=9740#step:7:6381).

File: engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java (hunk 1 above)
Comment (GregoryTravis) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: What was the test case that was fixed by f39794796c7f185c732fc08247d187282a58a24d?

Comment (GregoryTravis) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: I am curious, because this hash-caching technique might be a problem if we create the same `Decimal` value from a string, without the benefit of a `Number` to get a hash value from. This might allow two Decimals of the same value, one with a cached hash and one without. These values would be `==` but have different hashes.

I do like the idea you mentioned of converting everything to a double, or something like that -- that might make this unnecessary.

Comment (JaroslavTulach) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: > What was the test case that was fixed by [f397947](https://github.com/enso-org/enso/commit/f39794796c7f185c732fc08247d187282a58a24d)?

That fixes various [AssertionError](https://github.com/enso-org/enso/actions/runs/8749250691/job/24010598281?pr=9740#step:7:1209)s - if there is a `from` conversion, then the newly created atom must have the same hashcode as the old value.

> I am curious, because this hash-caching technique might be a problem if we create the same `Decimal` value from a string, without the benefit of a `Number` to get a hash value from. 

Yes, those would be different values when hashed and they should technically _not be equal_.

> This might allow two Decimals of the same value, one with a cached hash and one without. These values would be `==` but have different hashes.

Alas, that is true...

> I do like the idea you mentioned of converting everything to a double, or something like that -- that might make this unnecessary.

...OK, I'll try it. It seems like the only way out of the current misery.

Comment (JaroslavTulach) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: Here is my attempt to use `Double.hashCode` for hashing: https://github.com/enso-org/enso/pull/9740/commits/867262faedf6c7fa4fa6c5b066025a7b9506a526 It requires the `Decimal.from (that : Float)` **conversion to be exact** - e.g. `new BigDecimal(value)`- without that we'd get various [AssertionError](https://github.com/enso-org/enso/actions/runs/8749250691/job/24010598281?pr=9740#step:7:1209)s back. These are the currently failing tests (mostly result of the exact conversion, I think):
```
[FAILED] min/max: [2/3, 320ms]
    - should calculate min and max correctly [141ms]
    - [FAILED] should calculate min and max correctly, with mixed Decimal and Int/Float [172ms]
        Reason: 12.1 did not equal 12.1 (at test/Base_Tests/src/Data/Decimal_Spec.enso:929:13-79).
    - should calculate min and max correctly, with Math.min/max [7ms]
--
[FAILED] comparison: [3/4, 335ms]
    - should compare correctly [185ms]
    - should compare correctly, even with different internal representations [44ms]
    - [FAILED] should compare correctly to Integer and Float [94ms]
        Reason: (Decimal.Value 0.1) did not equal 0.1 (at test/Base_Tests/src/Data/Decimal_Spec.enso:307:17-41).
    - should not implicitly convert Decimal to Integer in comparison [11ms]
[FAILED] arithmetic: [6/7, 278ms]
    - should allow arithmetic with Decimals, without Math_Context [113ms]
    - should allow arithmetic with Decimals, with precision setting (examples) [15ms]
--
    - [FAILED] should negate values correctly [16ms]
        Reason: (Decimal.Value -5.3) did not equal -5.3 (at test/Base_Tests/src/Data/Decimal_Spec.enso:510:13-56).
read_text: [6/6, 93ms]
--
[FAILED] text conversion: [3/4, 20ms]
    - should convert to text correctly [4ms]
    - should convert to text without scientific notation correctly [2ms]
--
    - [FAILED] should parse correctly [3ms]
        Reason: (Decimal.Value 123456789.87654) did not equal 1.2345678987654E8 (at test/Base_Tests/src/Data/Decimal_Spec.enso:717:13-74).
should correctly evaluate documentation examples: [8/8, 43ms]
--
[FAILED] Asserts: [4/5, 6ms]
    - [FAILED] should be enabled in tests [3ms]
        Reason: Expected a value of type Standard.Base.Errors.Common.Assertion_Error but got a value of type Standard.Base.Nothing.Nothing instead (at test/Base_Tests/src/Runtime/Asserts_Spec.enso:16:9-52).
    - should be able to take foreign functions as expressions [0ms]
--
[FAILED] conversions: [7/8, 215ms]
    - should convert correctly to and from Integer [35ms]
    - should convert correctly to Float [37ms]
--
    - [FAILED] Decimal.to_float should compare correctly with the original Decimal [7ms]
        Reason: Expected False to be True (at test/Base_Tests/src/Data/Decimal_Spec.enso:594:13-53).
    - Decimal.to_integer should compare correctly with the original Decimal [4ms]
--
[FAILED] construction: [17/19, 135ms]
    - should be able to construct a Decimal from a string [2ms]
    - should throw Number_Parse_Error on a badly-formatted string [4ms]
--
    - [FAILED] should be able to explicity and implicitly specify infinite precision [5ms]
        Reason: (Decimal.Value 123.13) did not equal 123.13 (at test/Base_Tests/src/Data/Decimal_Spec.enso:83:13-76).
    - should report precision loss accurately [56ms]
--
    - [FAILED] should be convertible via .from [7ms]
        Reason: (Decimal.Value 123.4500000000000028421709430404007434844970703125) did not equal (Decimal.Value 123.45) (at test/Base_Tests/src/Data/Decimal_Spec.enso:120:13-67).
    - constructor should respect Math_Context  [7ms]
--
[FAILED] signs: [1/3, 7ms]
    - [FAILED] should calculate abs correctly [3ms]
        Reason: (Decimal.Value 12.345) did not equal 12.345 (at test/Base_Tests/src/Data/Decimal_Spec.enso:728:13-60).
    - [FAILED] should calculate negate correctly [1ms]
        Reason: (Decimal.Value -12.345) did not equal -12.345 (at test/Base_Tests/src/Data/Decimal_Spec.enso:734:13-64).
    - should calculate signum correctly [1ms]
--
[FAILED] remainder: [2/3, 40ms]
    - [FAILED] should define remainder [37ms]
        Reason: (Decimal.Value 0.2) did not equal 0.2 (at test/Base_Tests/src/Data/Decimal_Spec.enso:624:17-67).
    - should define remainder (example) [0ms]
--
[FAILED] rounding: [16/22, 162ms]
    - Can round positive decimals correctly [27ms]
    - Can round negative decimals correctly [13ms]
--
    - [FAILED] Can round positive decimals to a specified number of decimal places [4ms]
        Reason: (Decimal.Value 3.14) did not equal 3.14 (at test/Base_Tests/src/Data/Round_Spec.enso:42:9-46).
    - [FAILED] Can round negative decimals to a specified number of decimal places [4ms]
        Reason: (Decimal.Value -3.14) did not equal -3.14 (at test/Base_Tests/src/Data/Round_Spec.enso:55:9-48).
    - Can round positive decimals to a specified negative number of decimal places [17ms]
--
    - [FAILED] Banker's rounding handles half-way values correctly [13ms]
        Reason: (Decimal.Value 0.24) did not equal 0.24 (at test/Base_Tests/src/Data/Round_Spec.enso:106:9-62).
    - Banker's rounding handles non-half-way values just like normal rounding [19ms]
    - [FAILED] Can round correctly near the precision limit [4ms]
        Reason: (Decimal.Value 1.2222222223) did not equal 1.2222222223 (at test/Base_Tests/src/Data/Round_Spec.enso:132:9-62).
    - [FAILED] Can round correctly near the precision limit, using banker's rounding [5ms]
        Reason: (Decimal.Value 1.2222222222) did not equal 1.2222222222 (at test/Base_Tests/src/Data/Round_Spec.enso:157:9-79).
    - [FAILED] Floating point imperfect representation counter-examples [4ms]
        Reason: (Decimal.Value 1.22) did not equal 1.22 (at test/Base_Tests/src/Data/Round_Spec.enso:182:9-62).
    - Can round small integers to a specified number of decimal places correctly (value is unchanged) [2ms]
```

Comment (JaroslavTulach) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: Maybe we should [attach a warning](https://github.com/enso-org/enso/pull/9740/commits/9619afe8d16a4aaeb3694144fcc754eb8ed1e54c) doing no-parameter `Decimal.from (that : Float)` conversion. The conversion may be _too precise_ for regular user expectation!?

Comment (radeusgd) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: > Maybe we should [attach a warning](https://github.com/enso-org/enso/pull/9740/commits/9619afe8d16a4aaeb3694144fcc754eb8ed1e54c) doing no-parameter `Decimal.from (that : Float)` conversion. The conversion may be _too precise_ for regular user expectation!?

That sounds like a good idea

Comment (GregoryTravis) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: At least some of these test failures can be fixed by changing them from

```
            12.1 . min (Decimal.new "12.3") . should_equal (Decimal.new "12.1")
```
to
```
            12.1 . min (Decimal.new "12.3") . should_equal (Decimal.new 12.1)
```
(And using exact conversions everywhere.)

This would mean that `Decimal.new 12.1 != Decimal.new "12.1")`, which is strange, but is also what serious users of Java Decimal expect -- it is documented that the string argument is the preferred one.

What is *not* clear is how this is going to work with the database tables.

I will try this out and see how it goes.

Comment (radeusgd) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: As discussed on the standup, I think it may make sense to go for the _exact_ conversion by default (so that hashing will use it), but we can use the _'optional arguments of `.from`'_ to allow the user to also opt-in for the more familiar conversion.

Like
```
Decimal.from (that:Float) (exact:Boolean=True) = ...
```

**Comment by JaroslavTulach** [comment]
Prior to [8f475d1](https://github.com/enso-org/enso/pull/9740/commits/8f475d1fbc0587d7941711baf3d8df83022ab51d) we seem to have
- 17 failed tests in [Base_Tests](https://github.com/enso-org/enso/actions/runs/8810050813/job/24181812531?pr=9740#step:7:4267)
- 6 failed tests in [Table_Tests](https://github.com/enso-org/enso/actions/runs/8810050813/job/24181812531?pr=9740#step:7:11481)

Let's see what [8f475d1](https://github.com/enso-org/enso/pull/9740/commits/8f475d1fbc0587d7941711baf3d8df83022ab51d) does with the results. After the fix we have:
- 4 failed tests in [Base Tests](https://github.com/enso-org/enso/actions/runs/8810353135/job/24182586103?pr=9740#step:7:4253)
- 6 failed tests in [Table_Tests](https://github.com/enso-org/enso/actions/runs/8810353135/job/24182586103?pr=9740#step:7:4253)

Failures in `Table_Tests` might be fixed by [e5f21cf](https://github.com/enso-org/enso/pull/9740/commits/e5f21cf4c76c6de952c8d2b3962380b9e4b844f1) and `Base_Tests` shall be addressed by [d2383b7](https://github.com/enso-org/enso/pull/9740/commits/d2383b737c2784d75fd117af1eba9cf5abe1a053). Looks like the `should_equal` tests are now OK.

File: engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:376]: Right now (after [8f475d1](https://github.com/enso-org/enso/pull/9740/commits/8f475d1fbc0587d7941711baf3d8df83022ab51d)) it is:
- `Decimal.from 3.14` - exact precision
- `Decimal.new 3.14` - non-exact precision

e.g. it is _not necessary_ to add the `exact` argument to `from` conversion as we can choose the precision by calling either `from` or `new`.

**Comment by JaroslavTulach** [comment]
There is [one remaining failure](https://github.com/enso-org/enso/actions/runs/8812655536/job/24188761908?pr=9740#step:7:4055):
```
Reason: An unexpected panic was thrown: (No_Such_Method.Error Error UnresolvedSymbol<==>)
at <enso> Error.is_a<arg-0>(enso/built-distribution/enso-engine-*/enso-*/lib/Standard/Base/2024.1.1-dev/src/Meta.enso:294:28-35)
at <enso> Error.is_a(enso/built-distribution/enso-engine-*/enso-*/lib/Standard/Base/2024.1.1-dev/src/Meta.enso:294:28-54)
at <enso> <anonymous><arg-0>enso/test/Base_Tests/src/Semantic/Meta_Spec.enso:129:13-26)
```
probably unrelated, but it seems persistent. Worths investigation.

**Comment by radeusgd** [comment]
> There is [one remaining failure](https://github.com/enso-org/enso/actions/runs/8812655536/job/24188761908?pr=9740#step:7:4055):
> 
> ```
> Reason: An unexpected panic was thrown: (No_Such_Method.Error Error UnresolvedSymbol<==>)
> at <enso> Error.is_a<arg-0>(enso/built-distribution/enso-engine-*/enso-*/lib/Standard/Base/2024.1.1-dev/src/Meta.enso:294:28-35)
> at <enso> Error.is_a(enso/built-distribution/enso-engine-*/enso-*/lib/Standard/Base/2024.1.1-dev/src/Meta.enso:294:28-54)
> at <enso> <anonymous><arg-0>enso/test/Base_Tests/src/Semantic/Meta_Spec.enso:129:13-26)
> ```
> 
> probably unrelated, but it seems persistent. Worths investigation.

This looks like the hidden issue that is being investigated: https://github.com/enso-org/enso/pull/9763 (cc: @hubertp)

**Comment by JaroslavTulach** [comment]
The [last failure](https://github.com/enso-org/enso/actions/runs/8812655536/job/24199543153?pr=9740#step:7:5331) is:
```
Execution finished with an error: An IO error has occurred: java.nio.file.FileSystemException: C:\runner\_work\enso\enso\test\Table_Tests\data\transient\out6.xls: The process cannot access the file because it is being used by another process (C:\runner\_work\enso\enso\test\Table_Tests\data\transient\out6.xls).
 INFO ide_ci::program::command: cmdℹ️         at <enso> Panic.throw(Internal)
 INFO ide_ci::program::command: cmdℹ️         at <enso> case_branch(C:\runner\_work\enso\enso\built-distribution\enso-engine-2024.1.1-dev-windows-amd64\enso-2024.1.1-dev\lib\Standard\Base\2024.1.1-dev\src\Error.enso:138:25-43)
 INFO ide_ci::program::command: cmdℹ️         at <enso> <anonymous>(C:\runner\_work\enso\enso\built-distribution\enso-engine-2024.1.1-dev-windows-amd64\enso-2024.1.1-dev\lib\Standard\Base\2024.1.1-dev\src\Error.enso:137-142)
 INFO ide_ci::program::command: cmdℹ️         at <enso> Error.catch_primitive(Internal)
 INFO ide_ci::program::command: cmdℹ️         at <enso> Error.catch(C:\runner\_work\enso\enso\built-distribution\enso-engine-2024.1.1-dev-windows-amd64\enso-2024.1.1-dev\lib\Standard\Base\2024.1.1-dev\src\Error.enso:136-142)
 INFO ide_ci::program::command: cmdℹ️         at <enso> Panic.type.rethrow(C:\runner\_work\enso\enso\built-distribution\enso-engine-2024.1.1-dev-windows-amd64\enso-2024.1.1-dev\lib\Standard\Base\2024.1.1-dev\src\Panic.enso:88:21-47)
```

**Comment by JaroslavTulach** [comment]
Running benchmarks:
- [engine](https://github.com/enso-org/enso/actions/runs/8817113454)
- [stdlib](https://github.com/enso-org/enso/actions/runs/8817117205)

The results of engine benchmarks are OKeyish. Neither better neither outside of scope of other results. The standard library benchmarks are (sometimes) improving. The key `Enso_Hash_Map` benchmark is actually getting better...

Hash_Map

The simplification of the `HashCodeNode` probably pays off - less instructions, less if statements - faster code.

**Review by GregoryTravis — APPROVED** [review]

**Comment by JaroslavTulach** [comment]
Another promising benchmark result:

image

`Vector.distinct` on `Integer` seems to also reach fourteen days low - good:

Vector.distinct

Btw. `Vector.distinct` on `Text` remains unchanged. Looks like the `HashCodeNode` simplification really pays off.
